### PR TITLE
fix: 🐛 do not remove self parameter annotation when types do not match

### PR DESCRIPTION
**Pull request by ringohoffman**
Always removing the first parameter's type annotation when its name is `"self"` can mask underlying issues in the C++ binding, which is more obviously revealed when you only remove the parameter's annotation when it matches the class name it is bound to or it is `typing.Any`.

I noticed this bug since there was an unused import in `demo/_bindings/aliases/foreign_method_arg.pyi`, which seemed strange until I noticed that the sole parameter was actually in fact typed as a `Foo` in the C++ code (the C++ method implementation is missing a `self` parameter). I think making this change will help to make these bugs in the underlying C++ code more obvious:

Screenshot 2023-11-23 at 5 55 31 PM

### CODE REVIEW

File: tests/stubs/python-3.12/pybind11-master/numpy-array-use-type-var/demo/_bindings/aliases/foreign_method_arg.pyi [hunk 1]
```diff
@@ -5,4 +5,4 @@ import demo._bindings.classes
 __all__ = ["Bar2"]
 
 class Bar2:
-    def set_foo(self) -> int: ...
+    def set_foo(self: demo._bindings.classes.Foo) -> int: ...
```
Comment (sizmailov) [inline tests/stubs/python-3.12/pybind11-master/numpy-array-use-type-var/demo/_bindings/aliases/foreign_method_arg.pyi:8]: I believe it's a bug in `demo` module. Let's fix it and have a separate set of tests for this PR.

File: pybind11_stubgen/parser/mixins/fix.py [hunk 2]
```diff
@@ -43,22 +43,35 @@ class RemoveSelfAnnotation(IParser):
...
+        path = QualifiedName(path[:-1])  # remove method name from path
+        first_arg = func.args[0]
+        if (
+            first_arg.name == "self"
+            and isinstance(first_arg.annotation, ResolvedType)
+            and (
+                first_arg.annotation.name == QualifiedName.from_str("Any")
+                or first_arg.annotation.name == QualifiedName.from_str("typing.Any")
+                or first_arg.annotation.name == path
+                or first_arg.annotation.name == path[-len(first_arg.annotation.name) :]
```
Comment (sizmailov) [inline pybind11_stubgen/parser/mixins/fix.py:71]: let's change it to smth like `first_arg.annotation.name in {any_t_name, full_any_t_name, self_class_name, full_self_class_name}`

Let's also check for empty parameters of `ResolvedType`

**Review by sizmailov — CHANGES_REQUESTED** [review]

**Comment by sizmailov** [comment]
I thought that it would be nice to produce a warning when we find a suspicious `self` annotation. 
If the problem is reliably detectable by static analysis checkers, we don't need to.

File: pybind11_stubgen/parser/mixins/fix.py (hunk 2 above)
Comment (ringohoffman) [reply pybind11_stubgen/parser/mixins/fix.py:71]: Would you be interested in turning all of the `QualifiedName.from_str(<static string>)` into constants at the top of this file?

Comment (ringohoffman) [reply pybind11_stubgen/parser/mixins/fix.py:71]: Is it preferable to do `ResolvedType(name=TYPING_ANY_T_NAME)` than `self.parse_annotation_str("Any")`? I think the obvious benefit is reusing the constant. What do you think?

Comment (sizmailov) [reply pybind11_stubgen/parser/mixins/fix.py:71]: Let's make them class private constants like in other "fix" classes.

**Review by sizmailov — APPROVED** [review]

**Pull request merged**